src/requests.rs

```diff
@@ -423,16 +410,96 @@ macro_rules! request {
 request!(KeysBackupRequest from OriginalKeysBackupRequest extracts version: string and groups rooms);
 
-// Other Requests
-request!(SigningKeysUploadRequest from OriginalUploadSigningKeysRequest groups master_key, self_signing_key, user_signing_key);
+// Specific conversion for signature upload as they can be returned directly by
+// some verification API and not via outgoing requests.
+impl TryFrom<&OriginalSignatureUploadRequest> for SignatureUploadRequest {
+    type Error = serde_json::Error;
+    fn try_from(request: &OriginalSignatureUploadRequest) -> Result<Self, Self::Error> {
+        {
+            Ok(SignatureUploadRequest {
+                id: None,
+                signed_keys: serde_json::to_string(&request.signed_keys)?.into(),
+            })
+        }
+    }
+}
+
+// Not handled by macro as there is no request_id.
+impl TryFrom<&OriginalUploadSigningKeysRequest> for SigningKeysUploadRequest {
+    type Error = serde_json::Error;
+    fn try_from(request: &OriginalUploadSigningKeysRequest) -> Result<Self, Self::Error> {
+        {
+            Ok(SigningKeysUploadRequest {
+                body: {
+                    let mut map = serde_json::Map::new();
+                    map.insert(
+                        "master_key".to_owned(),
+                        serde_json::to_value(&request.master_key).unwrap(),
+                    );
+                    map.insert(
+                        "self_signing_key".to_owned(),
+                        serde_json::to_value(&request.self_signing_key).unwrap(),
+                    );
+                    map.insert(
+                        "user_signing_key".to_owned(),
+                        serde_json::to_value(&request.user_signing_key).unwrap(),
+                    );
+                    let object = serde_json::Value::Object(map);
+                    serde_json::to_string(&object)?.into()
+                },
+            })
+        }
+    }
+}
+
+// Manual into/from for signature upload requests as the id is optional.
+impl TryFrom<(String, &OriginalSignatureUploadRequest)> for SignatureUploadRequest {
+    type Error = serde_json::Error;
+    fn try_from(
+        (request_id, request): (String, &OriginalSignatureUploadRequest),
+    ) -> Result<Self, Self::Error> {
+        {
+            Ok(SignatureUploadRequest {
+                id: Some(request_id.into()),
+                signed_keys: serde_json::to_string(&request.signed_keys)?.into(),
+            })
+        }
+    }
+}
+
+// Specific conversion for to-device as they can be returned directly instead of
+// via outgoing requests. See `share_room_key`, the request_id is the txn_id.
+impl TryFrom<&OriginalToDeviceRequest> for ToDeviceRequest {
+    type Error = serde_json::Error;
+    fn try_from(request: &OriginalToDeviceRequest) -> Result<Self, Self::Error> {
+        {
+            Ok(ToDeviceRequest {
+                id: request.txn_id.to_string().into(),
```

Comment:
```suggestion
                // When matrix_sdk_crypto returns the request via `share_room_key`, it 
                // also stores the request as an "outgoing request" for future return
                // via `OlmMachine::outgoing_requests`, using the `txn_id` of the to-device
                // request as the "request ID" of the outgoing request.
                //
                // We therefore set `id = txn_id` here. Doing so means that the application
                // code will `mark_request_as_sent` once it has sent the request, passing the
                // txn ID back as the request ID, correctly marking the request as sent.
                id: request.txn_id.to_string().into(),
```

Reply:
I did some additional check and actually room keys to-device are never returned by `OlmMachine::outgoing_requests`, they are kept separatly in store attached to the outbound session.
So they'll be return again by the next call to share_room_key in this room

Reply:
I created an issue to update the rust sdk docu https://github.com/matrix-org/matrix-rust-sdk/issues/2738